### ISSUE TITLE
com.chaquo.python.PyException: FileExistsError: /data/data/org.electroncash.wallet/files/data/wallets/4

### ISSUE THREAD

**Issue by ProfessorVenkman** [Electron-Cash/Electron-Cash#1685]
Crash Report
============

This crash report was reported through the automatic crash reporting system 🤖

Traceback
--------------

```Python traceback
	at <python>.electroncash_gui.android.console.create(console.py:137)
	at <python>.chaquopy_java.call(chaquopy_java.pyx:283)
	at <python>.chaquopy_java.Java_com_chaquo_python_PyObject_callAttrThrows(chaquopy_java.pyx:255)
	at com.chaquo.python.PyObject.callAttrThrows(Native Method)
	at com.chaquo.python.PyObject.callAttr(PyObject.java:207)
	at org.electroncash.electroncash3.NewWalletImportDialog.onCreateWallet(NewWallet.kt:195)
	at org.electroncash.electroncash3.NewWalletDialog2.doInBackground(NewWallet.kt:101)
	at org.electroncash.electroncash3.NewWalletDialog2.doInBackground(NewWallet.kt:84)
	at org.electroncash.electroncash3.LaunchedTaskDialog.doInBackground(Dialog.kt:244)
	at org.electroncash.electroncash3.TaskDialog$onStart$1.run(Dialog.kt:184)
	at java.lang.Thread.run(Thread.java:764)
com.chaquo.python.PyException: FileExistsError: /data/data/org.electroncash.wallet/files/data/wallets/4
```

Reporter
------------

This issue was reported by 1 user(s):

| Electron Cash Version  | Python Version | Operating System  | Wallet Type  | Locale |
|---|---|---|---|---|
| 4.0.11-0  | 3.6.5 (default, Sep 11 2019, 15:39:45)  [GCC 4.9] | Android 9 on samsung SM-G975F (beyond2) | unknown | in_ID | 

Additional Information
------------------------

The reporting user(s) did not provide additional information.

**Comment by cculianu** [comment]
Hmm @mhsmith ..  I feel like this is a dupe of an older issue but I can't find it now..

**Comment by mhsmith** [comment]
Looks like #1497, but that should already have been fixed in 4.0.11. I'll investigate.

**Comment by mhsmith** [comment]
For future reference, here's how this happened:

Clicking the OK button at the end of the wallet creation process displays the TaskDialog, which creates the wallet file on a background thread. If you click the button twice in quick succession, it's possible to queue two clicks before the listener runs. But as long as the only thing the listener does is call showDialog, that should be fine because showDialog ignores any attempt to show a dialog that already exists. 

Unfortunately, because showDialog was showing the dialog using an asynchronous fragment transaction (DialogFragment.show), the dialog still didn't exist the second time it was called, and it went ahead and queued another transaction, which means we got two dialogs, and two threads, both trying to create the same wallet file.

Fixed by making showDialog use a synchronous fragment transaction (DialogFragment.showNow). However, because transactions can't nest, that means showDialog is no longer re-entrant (#1850, #1860, #1875). I fixed those issues by refactoring the code so that the second dialog is now created by the first dialog's caller. It could also have been fixed by creating the second dialog from a runnable posted to the event queue, or by giving showDialog an option to use an asynchronous transaction.
